### PR TITLE
[WIP] [DO NOT MERGE] Keyword detection.

### PR DESCRIPTION
**Pull request by lgirdwood**
enum and mux updates for keyword detection. Untested atm. @ranj063 fyi

### CODE REVIEW

File: sound/soc/sof/control.c [hunk 1]
```diff
@@ -109,6 +109,150 @@ int snd_sof_volume_put(struct snd_kcontrol *kcontrol,
...
+	if (ret < 0) {
+		dev_err_ratelimited(sdev->dev, "error: enum get failed to resume %d\n",
+				    ret);
+		return ret;
+	}
+
+	/* get all the mixer data from DSP */
+	snd_sof_ipc_get_comp_data(sdev->ipc, scontrol, SOF_IPC_COMP_GET_VALUE,
+				  SOF_CTRL_TYPE_VALUE_CHAN_GET,
+				  SOF_CTRL_CMD_VOLUME);
```
Comment (ranj063) [inline sound/soc/sof/control.c:205]: @lgirdwood should this be SOF_CTLR_CMD_ENUM?

**Comment by ranj063** [comment]
@lgirdwood overall the changes look good. I think we just need some minor modifications to the ctrl cmd's. I will test this out tomorrow and post any incremental patches needed.

**Comment by lgirdwood** [comment]
@ranj063 ok, thanks - I'm hoping it's just minor changes needed. But please feel free to send a new PR and close this one once it's working.

**Comment by lgirdwood** [comment]
@ranj063 @plbossart here are some more updates pushed for PM D0i3 stream support. It's all just compile tested infrastructure atm as I wont be able to test on HW this week. 
Missing is the async/block forever stream support and setting the wake source.

File: sound/soc/soc-topology.c [hunk 2]
```diff
@@ -1364,6 +1364,10 @@ static struct snd_kcontrol_new *soc_tplg_dapm_widget_denum_create(
 
 	for (i = 0; i < num_kcontrols; i++) {
 		ec = (struct snd_soc_tplg_enum_control *)tplg->pos;
+
+		tplg->pos += (sizeof(struct snd_soc_tplg_enum_control) +
+				ec->priv.size);
+
```
Comment (plbossart) [inline sound/soc/soc-topology.c:1370]: looks like the position handling is inconsistent in soc-topology. it's sometimes done before the error checks and sometimes done after. I'd prefer the latter but don't quite get how the code was written in the first place

File: include/sound/sof/topology.h [hunk 3]
```diff
@@ -34,6 +34,8 @@ enum sof_comp_type {
 	SOF_COMP_EQ_FIR,
 	SOF_COMP_FILEREAD,	/**< host test based file IO */
 	SOF_COMP_FILEWRITE,	/**< host test based file IO */
+	SOF_COMP_KEYWORD_DETECT,
+	SOF_COMP_VOICE_DETECT,
```
Comment (plbossart) [inline include/sound/sof/topology.h:38]: should we move the FILEREAD/WRITE to e.g 1000 to keep a linear list for real components?

File: sound/soc/sof/control.c [hunk 4]
```diff
@@ -109,6 +109,150 @@ int snd_sof_volume_put(struct snd_kcontrol *kcontrol,
...
+	if (ret < 0) {
+		dev_err_ratelimited(sdev->dev, "error: volume get failed to resume %d\n",
+				    ret);
+		return ret;
+	}
+
+	/* get all the mixer data from DSP */
+	snd_sof_ipc_get_comp_data(sdev->ipc, scontrol, SOF_IPC_COMP_GET_VALUE,
+				  SOF_CTRL_TYPE_VALUE_CHAN_GET,
+				  SOF_CTRL_CMD_VOLUME);
```
Comment (plbossart) [inline sound/soc/sof/control.c:133]: a switch isn't necessarily related to volume control?

File: sound/soc/sof/control.c [hunk 5]
```diff
@@ -109,6 +109,150 @@ int snd_sof_volume_put(struct snd_kcontrol *kcontrol,
...
+	/* update each channel */
+	for (i = 0; i < channels; i++) {
+		cdata->chanv[i].value = ucontrol->value.integer.value[i];
+		cdata->chanv[i].channel = i;
+	}
+
+	/* notify DSP of mixer updates */
+	snd_sof_ipc_set_comp_data(sdev->ipc, scontrol, SOF_IPC_COMP_SET_VALUE,
+				  SOF_CTRL_TYPE_VALUE_CHAN_GET,
+				  SOF_CTRL_CMD_VOLUME);
```
Comment (plbossart) [inline sound/soc/sof/control.c:174]: same, why use volume here?

File: sound/soc/sof/control.c [hunk 6]
```diff
@@ -109,6 +109,150 @@ int snd_sof_volume_put(struct snd_kcontrol *kcontrol,
...
```
Comment (plbossart) [inline sound/soc/sof/control.c:246]: should be enum?

File: sound/soc/sof/topology.c [hunk 7]
```diff
@@ -1755,6 +1837,8 @@ static int sof_widget_ready(struct snd_soc_component *scomp, int index,
 		break;
 	case snd_soc_dapm_mux:
 	case snd_soc_dapm_demux:
+		ret = sof_widget_load_mux(scomp, index, swidget, tw, &reply);
+		break;
 	case snd_soc_dapm_switch:
```
Comment (plbossart) [inline sound/soc/sof/topology.c:1873]: load_switch?

File: sound/soc/sof/topology.c [hunk 8]
```diff
@@ -2713,6 +2797,8 @@ static int sof_manifest(struct snd_soc_component *scomp, int index,
 static const struct snd_soc_tplg_kcontrol_ops sof_io_ops[] = {
 	{SOF_TPLG_KCTL_VOL_ID, snd_sof_volume_get, snd_sof_volume_put},
 	{SOF_TPLG_KCTL_BYTES_ID, snd_sof_bytes_get, snd_sof_bytes_put},
+	{SOF_TPLG_KCTL_ENUM_ID, snd_sof_enum_get, snd_sof_enum_put},
+	{SOF_TPLG_KCTL_SWITCH_ID, snd_sof_switch_get, snd_sof_switch_put},
```
Comment (plbossart) [inline sound/soc/sof/topology.c:2846]: how does the switch get loaded?

File: include/uapi/sound/sof/tokens.h [hunk 9]
```diff
@@ -94,6 +94,7 @@
 /* Tone */
 #define SOF_TKN_TONE_SAMPLE_RATE		800
 
-#define SOF_TKN_EFFECT_TYPE                     900
+/* Processing Components */
+#define SOF_TKN_PROCESS_TYPE                    900
```
Comment (plbossart) [inline include/uapi/sound/sof/tokens.h:101]: for backwards compatibility we probably want to have a conversion from effect to process?

File: sound/soc/sof/pcm.c [hunk 10]
```diff
@@ -241,7 +241,7 @@ static int sof_pcm_trigger(struct snd_pcm_substream *substream, int cmd)
 	struct snd_sof_pcm *spcm = rtd->private;
 	struct sof_ipc_stream stream;
 	struct sof_ipc_reply reply;
-	int ret = 0;
+	int ret = 0, permit_d0i3;
```
Comment (plbossart) [inline sound/soc/sof/pcm.c:244]: D0, D0i1, D3 are valid states
D0i3 isn't really defined anywhere. D0i2 in some cases but without audio.

File: sound/soc/sof/pcm.c [hunk 11]
```diff
@@ -325,6 +325,42 @@ static int sof_pcm_trigger(struct snd_pcm_substream *substream, int cmd)
...
+		ret = pm_runtime_put_autosuspend(sdev->dev);
+		if (ret < 0)
+			dev_err(sdev->dev, "error: pcm close failed to idle %d\n",
+				ret);
+		break;
+	case SNDRV_PCM_TRIGGER_PAUSE_PUSH:
+		/* fall through */
+	case SNDRV_PCM_TRIGGER_SUSPEND:
+		/* fallthrough */
+	case SNDRV_PCM_TRIGGER_STOP:
```
Comment (plbossart) [inline sound/soc/sof/pcm.c:356]: confusion here. You inverted the entry and exit in D3. Pause/suspend/stop should not result in a get_sync.

File: sound/soc/sof/pcm.c [hunk 12]
```diff
@@ -325,6 +325,42 @@ static int sof_pcm_trigger(struct snd_pcm_substream *substream, int cmd)
 	/* send IPC to the DSP */
 	ret = sof_ipc_tx_message(sdev->ipc, stream.hdr.cmd, &stream,
 				 sizeof(stream), &reply, sizeof(reply));
+	if (ret < 0)
+		return ret;
+
+	/* do we need to inform runtime PM we can enter D3 ? */
+	permit_d0i3 = snd_sof_pipeline_permit_d0i3(sdev,
+						   spcm->swidget->pipeline_id);
```
Comment (plbossart) [inline sound/soc/sof/pcm.c:333]: Is this correct? I thought the information that the dmic dailink is configured with .ignore_suspend was enough for DAPM/ALSA. @ranj063 can you comment?

File: sound/soc/sof/topology.c [hunk 13]
```diff
@@ -516,6 +533,16 @@ static const struct sof_topology_token pcm_tokens[] = {
...
 };
 
+/* Generic components PM */
+static const struct sof_topology_token pm_tokens[] = {
+	{SOF_TKN_PM_COMP_WAKE_SOURCE,
+		SND_SOC_TPLG_TUPLE_TYPE_WORD, get_token_u32,
+		offsetof(struct snd_sof_widget, pm_wake_source), 0},
+	{SOF_TKN_PM_COMP_PERMIT_D0I3,
+		SND_SOC_TPLG_TUPLE_TYPE_WORD, get_token_u32,
+		offsetof(struct snd_sof_widget, pm_permit_d0i3), 0},
```
Comment (plbossart) [inline sound/soc/sof/topology.c:543]: looks like these two properties are boolean?

File: include/uapi/sound/sof/tokens.h [hunk 14]
```diff
@@ -64,6 +64,8 @@
 #define SOF_TKN_COMP_FORMAT			402
 #define SOF_TKN_COMP_PRELOAD_COUNT		403
 #define	SOF_TKN_COMP_FLAVOUR			404
+#define	SOF_TKN_COMP_STREAM_ASYNC		405
+#define	SOF_TKN_COMP_STREAM_BLOCK		406
```
Comment (plbossart) [inline include/uapi/sound/sof/tokens.h:68]: @keyonjie can you comment on this one?

**Comment by ranj063** [comment]
@lgirdwood @plbossart I've got the enum/switch controls parts working now. I will submit a new PR for it. I will keep this PR open in the meanwhile for discussions and submit relevant parts as and when they are ready.

File: sound/soc/sof/control.c (hunk 6 above)
Comment (ranj063) [reply sound/soc/sof/control.c:246]: @plbossart i've fixed all this in my PR.

File: sound/soc/sof/pcm.c (hunk 12 above)
Comment (ranj063) [reply sound/soc/sof/pcm.c:333]: @plbossart @lgirdwood yes ignore_suspend is already defined for the DMIC dai link that is used for WoV in the machine driver. This is already used to determine if the stream should be suspended when the devices enters S0ix.

File: sound/soc/sof/topology.c [hunk 15]
```diff
@@ -183,6 +183,23 @@ static int set_up_volume_table(struct snd_sof_control *scontrol,
...
+	struct snd_sof_widget *swidget = NULL;
+
+	list_for_each_entry(swidget, &sdev->widget_list, list) {
+		if (swidget->pipeline_id != pipeline_id)
+			continue;
+		if (swidget->pm_permit_d0i3)
+			return 0;
+	}
+
+	return 0;
```
Comment (ranj063) [inline sound/soc/sof/topology.c:199]: @lgirdwood this  seems incorrect. Should we return 0 even if the pipeline is not found?

File: sound/soc/sof/pcm.c [hunk 16]
```diff
@@ -325,6 +325,42 @@ static int sof_pcm_trigger(struct snd_pcm_substream *substream, int cmd)
...
+	if (!permit_d0i3)
+		return 0;
+
+	/* inform runtime PM driver can enter D3 */
```
Comment (ranj063) [inline sound/soc/sof/pcm.c:337]: @plbossart @lgirdwood I'm a bit confused about this part. We will not even get a suspend trigger for streams associated with dai_links whose .ignore_suspend is set. This should help us determine when to activate D3.

**Comment by plbossart** [comment]
> > @@ -325,6 +325,42 @@ static int sof_pcm_trigger(struct snd_pcm_substream *substream, int cmd)
>   	/* send IPC to the DSP */
>   	ret = sof_ipc_tx_message(sdev->ipc, stream.hdr.cmd, &stream,
>   				 sizeof(stream), &reply, sizeof(reply));
> +	if (ret < 0)
> +		return ret;
> +
> +	/* do we need to inform runtime PM we can enter D3 ? */
> +	permit_d0i3 = snd_sof_pipeline_permit_d0i3(sdev,
> +						   spcm->swidget->pipeline_id);
> +	if (!permit_d0i3)
> +		return 0;
> +
> +	/* inform runtime PM driver can enter D3 */
>
> @plbossart <https://github.com/plbossart> @lgirdwood 
> <https://github.com/lgirdwood> I'm a bit confused about this part. We 
> will not even get a suspend trigger for streams associated with 
> dai_links whose .ignore_suspend is set. This should help us determine 
> when to activate D3.
>
I am really confused about the .ignore_suspend part. It's currently set 
for DMIC but that doesn't prevent us from suspending, does it?

**Comment by ranj063** [comment]
> > @@ -325,6 +325,42 @@ static int sof_pcm_trigger(struct snd_pcm_substream *substream, int cmd) /* send IPC to the DSP */ ret = sof_ipc_tx_message(sdev->ipc, stream.hdr.cmd, &stream, sizeof(stream), &reply, sizeof(reply)); +	if (ret < 0) +	return ret; + +	/* do we need to inform runtime PM we can enter D3 ? */ +	permit_d0i3 = snd_sof_pipeline_permit_d0i3(sdev, +	spcm->swidget->pipeline_id); +	if (!permit_d0i3) +	return 0; + +	/* inform runtime PM driver can enter D3 */ @plbossart <https://github.com/plbossart> @lgirdwood <https://github.com/lgirdwood> I'm a bit confused about this part. We will not even get a suspend trigger for streams associated with dai_links whose .ignore_suspend is set. This should help us determine when to activate D3.
> I am really confused about the .ignore_suspend part. It's currently set for DMIC but that doesn't prevent us from suspending, does it?

@plbossart yes it doesnt. because the PCM is not running/draining. if the PM is not running snd_soc_suspend() doesnt suspend it anyway

**Comment by plbossart** [comment]
@ranj063 this PR was replaced by all your other ones, right?

**Comment by ranj063** [comment]
> @ranj063 this PR was replaced by all your other ones, right?

@plbossart all but the last 2. I'll open a separate PR for those today and close this one right after.

**Comment by plbossart** [comment]
so are we good to close now?

**Comment by ranj063** [comment]
@plbossart yes I'll close this now. I didnt do a PR for the permit_D3 parts. I will defer them for later.

**Pull request closed without merging**